### PR TITLE
AWS: support Server Side Encryption for S3FileIO

### PR DESCRIPTION
**Pull request by jackye1995**
- add support for server side encryption (3 types: S3, KMS, Custom) for `S3FileIO`
- add corresponding integration tests
- add `AwsCatalogProperties` as the centralized place to put all AWS related catalog properties
- also fix warnings in `S3URI`

### CODE REVIEW

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 1]
```diff
@@ -91,9 +104,36 @@ public void close() throws IOException {
     try {
       stream.close();
 
-      s3.putObject(
-          PutObjectRequest.builder().bucket(location.bucket()).key(location.key()).build(),
-          RequestBody.fromFile(stagingFile));
+      PutObjectRequest.Builder requestBuilder = PutObjectRequest.builder()
```
Comment (rdsr) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:107]: Q: Why isn't the parent `super.close()` inside the try block as well?

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:107]: It's called on line 101

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:107]: Ah sorry misread the question. Technically `super.close()` calls `PositionOutputStream.close()` calls `OutputStream.close()` which does nothing, so I think it is fine to not have it in the try block.

File: aws/src/integration/java/org/apache/iceberg/aws/AwsIntegTestUtil.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  private AwsIntegTestUtil() {
+  }
+
+  /**
+   * Set the environment variable AWS_TEST_BUCKET for a default bucket to use for testing
+   * @return bucket name
+   */
+  public static String testBucketName() {
+    return System.getenv("AWS_TEST_BUCKET");
```
Comment (rdsr) [inline aws/src/integration/java/org/apache/iceberg/aws/AwsIntegTestUtil.java:40]: I assume these tests are not part of the regular build?

Comment (jackye1995) [reply aws/src/integration/java/org/apache/iceberg/aws/AwsIntegTestUtil.java:40]: No

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 3]
```diff
@@ -73,4 +89,19 @@ private S3Client client() {
     }
     return client;
   }
+
+  @Override
+  public void initialize(Map<String, String> properties) {
+    this.sseType = properties.getOrDefault(AwsCatalogProperties.S3_SSE_TYPE, AwsCatalogProperties.S3_SSE_TYPE_NONE);
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:95]: As mentioned in #1767, it would be nice to wrap these properties in AwsCatalogProperties as a utility class and just keep reference to the properties as opposed to enumerating them here an in BaseS3File.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3URI.java [hunk 4]
```diff
@@ -57,7 +57,7 @@
     Preconditions.checkNotNull(location, "Location cannot be null.");
 
     this.location = location;
-    String [] schemeSplit = location.split(SCHEME_DELIM);
+    String [] schemeSplit = location.split(SCHEME_DELIM, -1);
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3URI.java:60]: I'm trying to understand the reason for these (and below split changes).  Is this relevant to the larger PR?  My understanding is that this really doesn't change the behavior.

File: aws/src/main/java/org/apache/iceberg/aws/AwsCatalogProperties.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * Type of S3 Server side encryption used, default to {@link AwsCatalogProperties#S3_SSE_TYPE_NONE}.
+   */
+  public static final String S3_SSE_TYPE = "s3-sse-type";
+
+  /**
+   * No server side encryption.
+   */
+  public static final String S3_SSE_TYPE_NONE = "none";
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/AwsCatalogProperties.java:35]: These four types might make sense to turn into an enumeration.  Also, are these official names for the encryption types?  I just find `s3` not to be terribly descriptive for the type of encryption.

File: aws/src/main/java/org/apache/iceberg/aws/AwsCatalogProperties.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws;
+
+public class AwsCatalogProperties {
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/AwsCatalogProperties.java:22]: We might want to consider shortening this to just `AwsProperties` because while you can define them at the catalog level, tables can deviate and it would make sense to tune various properties for specific datasets within a catalog.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputFile.java [hunk 7]
```diff
@@ -29,7 +30,11 @@
 
 public class S3OutputFile extends BaseS3File implements OutputFile {
   public S3OutputFile(S3Client client, S3URI uri) {
-    super(client, uri);
+    this(client, uri, AwsCatalogProperties.S3_SSE_TYPE_NONE, null, null);
+  }
+
+  public S3OutputFile(S3Client client, S3URI uri, String sseType, String sseKey, String sseMd5) {
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputFile.java:36]: As I mentioned in the comment above, this is what I think we want to try to avoid where we have to add constructor params per property we need to add.  There are already a few others we need to push through (multipart size, canned acl, etc.).  I think turning the `AwsProperties` into a wrapper would help as we expand this functionality.

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputFile.java:36]: Yes I agree it sounds better to have all configuration in the separated class. Let me update it.

File: aws/src/main/java/org/apache/iceberg/aws/AwsCatalogProperties.java (hunk 5 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/AwsCatalogProperties.java]: Yeah, but unfortunately that's the official name. I am hesitated to turn them into enums, since it seems like in Iceberg people just use string instead of enum in many different places.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3URI.java (hunk 4 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3URI.java:60]: It's not, I am just fixing errorprone warnings along the way. Let me remove them since this PR seems pretty big at this point of time.

**Comment by danielcweeks** [comment]
@jackye1995 Thanks a lot for this!  +1
